File: engine/include/xjmusic/entities/meme/MemeConstellation.java
// Copyright (c) 1999-2022, XJ Music Inc. (https://xj.io) All Rights Reserved.

package io.xj.model.meme;

import java.util.std::vector;
import java.util.HashMap;
import java.util.std::vector;
import java.util.std::map;
import java.util.std::set;
import java.util.TreeSet;

/**
 Compute normalized string representation of an unordered set of memes
 for the purpose of identifying unique constellations.
 <p>
 for each unique sequence-pattern-meme constellation within the main sequence https://github.com/xjmusic/workstation/issues/208
 */
public enum MemeConstellation {
  ;
  static final std::string CONSTELLATION_DELIMITER = "_";

  /**
   @return unique constellation for this set of names
   */
  public static std::string fromNames(std::vector<std::string> names) {
    std::map<std::string, Boolean> uniqueNames = new HashMap<>();
    names.forEach(meme -> uniqueNames.put(meme, true));
    std::set<std::string> pieces = new TreeSet<>(uniqueNames.keySet());
    return std::string.join(CONSTELLATION_DELIMITER, pieces);
  }

  /**
   @return set of names parsed from constellation
   */
  public static std::vector<std::string> toNames(std::string constellation) {
    return std::vector.of(constellation.split(CONSTELLATION_DELIMITER));
  }
}

File: engine/src/entities/meme/MemeConstellation.cpp
// Copyright (c) XJ Music Inc. (https://xj.io) All Rights Reserved.

#include "xjmusic/entities/meme/MemeConstellation.h"

using namespace XJ;

std::string MemeConstellation::fromNames(const std::set<std::string> &names) {
  std::unordered_map<std::string, bool> uniqueNames;
  for (const auto &meme: names) {
    uniqueNames[meme] = true;
  }
  std::set<std::string> pieces;
  for (const auto &pair: uniqueNames) {
    pieces.insert(pair.first);
  }
  return join(CONSTELLATION_DELIMITER, pieces);
}

std::vector<std::string> MemeConstellation::toNames(const std::string &constellation)  {
  return split(constellation, CONSTELLATION_DELIMITER);
}

std::string MemeConstellation::join(const std::string &delimiter, const std::set<std::string> &pieces) {
  std::ostringstream joined;
  std::copy(pieces.begin(), pieces.end(), std::ostream_iterator<std::string>(joined, delimiter.c_str()));
  std::string result = joined.str();
  return result.substr(0, result.size() - delimiter.size()); // remove the last delimiter
}

std::vector<std::string> MemeConstellation::split(const std::string &str, const std::string &delimiter)  {
  std::vector<std::string> tokens;
  size_t start = 0, end;
  while ((end = str.find(delimiter, start)) != std::string::npos) {
    tokens.push_back(str.substr(start, end - start));
    start = end + delimiter.size();
  }
  tokens.push_back(str.substr(start));
  return tokens;
}
